### PR TITLE
ZK-3179: Toolbar can have an overflow button that contains those butto…

### PR DESCRIPTION
**Pull request by charlesqqq**
…ns weren't able to fit in the toolbar

### CODE REVIEW

File: zktest/test/java/org/zkoss/zktest/zats/test2/F86_ZK_3179Test.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+		waitResponse();
+		Assert.assertEquals(getToolbarChildrenLength(), 4);
+
+		click(overflowButton);
+		waitResponse();
+		Assert.assertEquals(getPopupChildrenLength(), 2);
+
+		click(findButtonByContent("setWidth(\"700px\")"));
+		waitResponse();
+		Assert.assertEquals(getToolbarChildrenLength(), 6);
```
Comment (wenninghsu) [inline zktest/test/java/org/zkoss/zktest/zats/test2/F86_ZK_3179Test.java]: do we need to test so many widths?

File: zul/src/archive/metainfo/xml/zul.xsd [hunk 2]
```diff
@@ -2858,6 +2858,7 @@
 		<xs:attributeGroup ref="xulElementAttrGroup" />
 		<xs:attribute name="orient" type="orientAttrType" use="optional" />
 		<xs:attribute name="align" type="packAttrType" use="optional"/>
+		<xs:attribute name="overflow" type="booleanType" use="optional"/>
```
Comment (wenninghsu) [inline zul/src/archive/metainfo/xml/zul.xsd]: overflow seems wierd, overflowMenu?

File: zul/src/archive/web/js/zul/wgt/Toolbar.js [hunk 3]
```diff
@@ -50,7 +50,72 @@ zul.wgt.Toolbar = zk.$extends(zul.Widget, {
...
+			if (child != this._overflowPopup)
+				children.push(child);
+			child = child.nextSibling;
+		}
+		return children;
+	},
+	_initOverflowPopup: function () {
+		var popup = this._overflowPopup = new zul.wgt.Popup();
+		popup.setSclass(this.$s('popup'));
+		this.appendChild(popup);
```
Comment (wenninghsu) [inline zul/src/archive/web/js/zul/wgt/Toolbar.js]: do we need to append it?

File: zul/src/archive/web/js/zul/wgt/Toolbar.js [hunk 4]
```diff
@@ -50,7 +50,72 @@ zul.wgt.Toolbar = zk.$extends(zul.Widget, {
...
+	},
+	_openPopup: function () {
+		this._overflowPopup.open(this, null, 'after_end');
```
Comment (wenninghsu) [inline zul/src/archive/web/js/zul/wgt/Toolbar.js]: if space not enough, what would it do?

File: zul/src/archive/web/js/zul/wgt/Toolbar.js [hunk 5]
```diff
@@ -94,5 +167,63 @@ zul.wgt.Toolbar = zk.$extends(zul.Widget, {
...
+
+		var toolbar = jq(this.$n()),
+			contentWidth = toolbar.width() - this._getPaddingWidth(),
+			initialChildrenWidth = this._initialChildrenWidth,
+			initialChildrenAmount = initialChildrenWidth.length,
+			tempChildrenWidth = 0,
+			newPopupChildrenAmount = 0;
+
+		for (var i = 0; i < initialChildrenAmount; i++) {
+			tempChildrenWidth += initialChildrenWidth[i];
```
Comment (wenninghsu) [inline zul/src/archive/web/js/zul/wgt/Toolbar.js]: why use initial width of children?

File: zul/src/archive/web/js/zul/wgt/Toolbar.js [hunk 6]
```diff
@@ -94,5 +167,63 @@ zul.wgt.Toolbar = zk.$extends(zul.Widget, {
...
+		var popup = this._overflowPopup,
+			overflowOn = this.$s('overflow-on'),
+			overflowOff = this.$s('overflow-off'),
+			oldToolbarChildren = this._getChildren(this);
+
+		if (newPopupChildrenAmount) {
+			toolbar.addClass(overflowOn).removeClass(overflowOff);
+			if (popupChildrenDiff > 0) {
+				for (var i = 0; i < popupChildrenDiff; i++)
+					popup.insertBefore(oldToolbarChildren.pop(), popup.firstChild);
```
Comment (wenninghsu) [inline zul/src/archive/web/js/zul/wgt/Toolbar.js]: why not just manipulate dom?

File: zul/src/archive/web/js/zul/wgt/Toolbar.js [hunk 7]
```diff
@@ -94,5 +167,63 @@ zul.wgt.Toolbar = zk.$extends(zul.Widget, {
...
+			this._adjustContent();
+	},
+	_adjustContent: function () {
+		if (zUtl.isImageLoading() || !this._initialChildrenWidth) {
+			setTimeout(this.proxy(this._adjustContent), 20);
+			return;
+		}
+
+		var toolbar = jq(this.$n()),
+			contentWidth = toolbar.width() - this._getPaddingWidth(),
```
Comment (wenninghsu) [inline zul/src/archive/web/js/zul/wgt/Toolbar.js]: why do we need to do - this._getPaddingWidth()

File: zul/src/archive/web/js/zul/wgt/less/toolbar.less [hunk 8]
```diff
@@ -6,6 +6,33 @@
 	padding: 6px 16px;
 	background: @baseBackgroundColor;
 	position: relative;
+
+	&-overflow {
+		white-space: nowrap;
+
+		&-on > .z-icon-ellipsis-v {
```
Comment (wenninghsu) [inline zul/src/archive/web/js/zul/wgt/less/toolbar.less]: should not use icon name directly

File: zul/src/archive/web/js/zul/wgt/mold/toolbar.js [hunk 9]
```diff
@@ -22,5 +24,7 @@ function (out) {
 		if ('horizontal' == this.getOrient() && zk.ie9)
 			out.push('<span></span>');
 	}
-	out.push('</div><div class="z-clear"></div></div>');
+	out.push('</div>');
+	out.push('<i id="', uuid, '-overflow" class="z-icon-ellipsis-v"/>');
```
Comment (wenninghsu) [inline zul/src/archive/web/js/zul/wgt/mold/toolbar.js]: should modify class name

File: zul/src/org/zkoss/zul/Toolbar.java [hunk 10]
```diff
@@ -105,11 +108,61 @@ public void setOrient(String orient) throws WrongValueException {
 		}
 	}
 
+	/**
+	 * Return whether toolbar has a overflow button that show a popup
```
Comment (wenninghsu) [inline zul/src/org/zkoss/zul/Toolbar.java]: an overflow button that shows a popup

File: zul/src/org/zkoss/zul/Toolbar.java [hunk 11]
```diff
@@ -105,11 +108,61 @@ public void setOrient(String orient) throws WrongValueException {
...
+	 * which contains those content weren't able to fit in the toolbar.
+	 * If overflow is false, toolbar will display multiple rows when content is wider than toolbar.
+	 * Default: false.
+	 *
+	 * @return whether toolbar has a overflow button
```
Comment (wenninghsu) [inline zul/src/org/zkoss/zul/Toolbar.java]: an overflow

File: zul/src/org/zkoss/zul/Toolbar.java [hunk 12]
```diff
@@ -105,11 +108,61 @@ public void setOrient(String orient) throws WrongValueException {
...
+		if (_overflow != overflow) {
+			_overflow = overflow;
+			smartUpdate("overflow", _overflow);
+
+			if (_overflow)
+				setAttribute(PRELOAD_IMAGE, "true");
+			else
+				removeAttribute(PRELOAD_IMAGE);
+
+			invalidate();
```
Comment (wenninghsu) [inline zul/src/org/zkoss/zul/Toolbar.java]: why use invalidate

File: zul/src/org/zkoss/zul/Toolbar.java [hunk 13]
```diff
@@ -105,11 +108,61 @@ public void setOrient(String orient) throws WrongValueException {
...
+		}
+	}
+
+	@Override
+	public void onChildAdded(Component child) {
+		super.onChildAdded(child);
+		if (isOverflow())
+			invalidate();
```
Comment (wenninghsu) [inline zul/src/org/zkoss/zul/Toolbar.java]: should avoid invalidate

File: zul/src/org/zkoss/zul/Toolbar.java [hunk 14]
```diff
@@ -105,11 +108,61 @@ public void setOrient(String orient) throws WrongValueException {
...
+	}
+
+	@Override
+	public void onChildRemoved(Component child) {
+		super.onChildRemoved(child);
+		if (isOverflow())
+			invalidate();
```
Comment (wenninghsu) [inline zul/src/org/zkoss/zul/Toolbar.java]: should avoid invalidate

File: zul/src/archive/web/js/zul/wgt/Toolbar.js [hunk 15]
```diff
@@ -50,7 +50,55 @@ zul.wgt.Toolbar = zk.$extends(zul.Widget, {
...
+		if (popup) {
+			popup.detach();
+			this.domUnlisten_(this.$n('overflowpopup-button'), 'onClick', '_openPopup');
+			zWatch.unlisten({onCommandReady: this});
+			zWatch.unlisten({onSize: this});
+		}
+		this.$supers('unbind_', arguments);
+	},
+	_initPopup: function () {
+		var popup = this._pp = new zul.wgt.Popup();
```
Comment (scribetw) [inline zul/src/archive/web/js/zul/wgt/Toolbar.js]: Just use div to build a popup.
There is no need to use Popup widget.

You need to sync position manually

File: zul/src/archive/web/js/zul/wgt/Toolbar.js [hunk 16]
```diff
@@ -84,6 +134,19 @@ zul.wgt.Toolbar = zk.$extends(zul.Widget, {
 	inPanelMold: function () {
 		return this._mold == 'panel';
 	},
+	appendChild: function (child) {
+		this.$supers('appendChild', arguments);
+		var popupCave = this._pp ? this._pp.$n('cave') : null;
```
Comment (scribetw) [inline zul/src/archive/web/js/zul/wgt/Toolbar.js]: if we use div, this._pp is a DOM node.

File: zul/src/archive/web/js/zul/wgt/mold/toolbar.js [hunk 17]
```diff
@@ -22,5 +24,12 @@ function (out) {
 		if ('horizontal' == this.getOrient() && zk.ie9)
 			out.push('<span></span>');
 	}
-	out.push('</div><div class="z-clear"></div></div>');
+	out.push('</div>');
+
+	if (this.isOverflowPopup()) {
+		out.push('<i id="', uuid, '-overflowpopup-button"',
+				' class="', this.$s('overflowpopup-button'), ' z-icon-ellipsis-h"/>');
+		out.push('<div id="', uuid, '-popup" class="', this.$s('popup'), ' ', this.$s('popup-close'), '"></div>');
```
Comment (scribetw) [inline zul/src/archive/web/js/zul/wgt/mold/toolbar.js]: `uuid, '-pp"`

File: zul/src/archive/web/js/zul/wgt/Toolbar.js [hunk 18]
```diff
@@ -50,7 +50,144 @@ zul.wgt.Toolbar = zk.$extends(zul.Widget, {
...
+		for (var i = 0; i < childrenAmount; i++) {
+			tempChildrenWidth += jq(children[i]).outerWidth(true);
+			if (tempChildrenWidth >= contentWidth) {
+				newPopupChildrenAmount = childrenAmount - i;
+				break;
+			}
+		}
+		// If overflow popup has more than 8 children,
+		// setting its height as the sum of the top eight popup children's heights to enable scroll bar.
+		if (newPopupChildrenAmount > 8) {
```
Comment (scribetw) [inline zul/src/archive/web/js/zul/wgt/Toolbar.js]: What if the the first children is `<div height="1000px;" />`?

Perhaps we need a fixed height (250 / 350?).
https://developer.mozilla.org/en-US/docs/Web/CSS/max-height

**Pull request merged**